lib/consensus/isaac.go

```diff
@@ -225,6 +225,52 @@ func (is *ISAAC) Vote(b ballot.Ballot) (isNew bool, err error) {
 	return
 }
 
+// RemoveOldBallots checkes that `blt` has valid confirmed and proposed time.
+// If it is invalid, it is removed.
+// And if the invalid ballot is make by itself,
+// return `needRenewal` = true for rebroadcasting EXP ballot
+func (is *ISAAC) RemoveOldBallots(blt ballot.Ballot) (needRenewal bool) {
+	runningRound, found := is.RunningRounds[blt.VotingBasis().Index()]
+	if !found {
+		// if RunningRound is not found, this ballot will be stopped.
+		return false
+	}
+	var roundVote *RoundVote
+	var err error
+	roundVote, err = runningRound.RoundVote(blt.Proposer())
+	if err != nil {
+		return false
+	}
+
+	switch blt.State() {
+	case ballot.StateSIGN:
+		for nodeAddr, blt := range roundVote.SIGN {
+			if err = ballot.CheckHasCorrectTime(blt.ProposerConfirmed()); err == nil {
+				if err = ballot.CheckHasCorrectTime(blt.Confirmed()); err == nil {
+					continue
+				}
+			}
+			delete(roundVote.SIGN, nodeAddr)
+			if nodeAddr == is.Node.Address() {
+				needRenewal = true
+			}
+		}
+	case ballot.StateACCEPT:
+		for nodeAddr, blt := range roundVote.ACCEPT {
```

Comment:
No way to pull this code together ? They have the same logic, just a different variables (`roundVote.{SIGN,ACCEPT}`)